wrappers/vroom.rb

```diff
@@ -29,175 +29,329 @@ def initialize(hash = {})
+        {
           id: index,
-          location_index: points[service.activity.point_id].matrix_index
-        }]
-      }.flatten
-
-      matrix_indices = (problem[:jobs].collect{ |jb| jb[:location_index] } + problem[:vehicles].collect{ |vec| [vec[:start_index], vec[:end_index]].uniq.compact }.flatten).uniq.sort
-      matrix = matrices.find{ |current_matrix| current_matrix.id == vehicle.matrix_id }
+          location_index: points[service.activity.point_id].matrix_index,
+          service: service.activity.duration.to_i,
+          skills: service.skills.flat_map{ |skill| vrp_skills.find_index{ |sk| sk == skill } }.compact + # undefined skills are ignored
+            service.sticky_vehicles.flat_map{ |sticky| vrp_skills.find_index{ |sk| sk == sticky.id } }.compact,
+          priority: (100 * (8 - service.priority).to_f / 8).to_i, # Scale from 0 to 100 (higher is more important)
+          time_windows: service.activity.timewindows.map{ |timewindow| [timewindow.start || 0, timewindow.end || 2**30] },
+          delivery: vrp_units.map{ |unit| service.quantities.find{ |quantity| quantity.unit.id == unit.id && quantity.value.negative? }&.value&.to_i || 0 },
+          pickup: vrp_units.map{ |unit| service.quantities.find{ |quantity| quantity.unit.id == unit.id && quantity.value.positive? }&.value&.to_i || 0 }
+        }.delete_if{ |k, v|
+          v.nil? || v.is_a?(Array) && v.empty? ||
+            k == :delivery && pickup_flag ||
+            k == :anykey && !pickup_flag
+        }
+      }
+      problem[:shipments] = vrp.shipments.map.with_index{ |shipment, index|
+        {
+          amount: vrp_units.map{ |unit| shipment.quantities.find{ |quantity| quantity.unit.id == unit.id && quantity.value&.positive? }&.value&.to_i || 0 },
+          skills: shipment.skills.flat_map{ |skill| vrp_skills.find_index{ |sk| sk == skill } }.compact + # undefined skills are ignored
+            shipment.sticky_vehicles.flat_map{ |sticky| vrp_skills.find_index{ |sk| sk == sticky.id } }.compact,
+          priority: (100 * (8 - shipment.priority).to_f / 8).to_i,
+          pickup: {
+            id: vrp.services.size + index * 2,
+            service: shipment.pickup.duration,
+            location_index: points[shipment.pickup.point_id].matrix_index,
+            time_windows: shipment.pickup.timewindows.map{ |timewindow| [timewindow.start || 0, timewindow.end || 2**30] }
+          }.delete_if{ |_k, v| v.nil? || v.is_a?(Array) && v.empty? },
+          delivery: {
+            id: vrp.services.size + index * 2 + 1,
+            service: shipment.delivery.duration,
+            location_index: points[shipment.delivery.point_id].matrix_index,
+            time_windows: shipment.delivery.timewindows.map{ |timewindow| [timewindow.start || 0, timewindow.end || 2**30] }
+          }.delete_if{ |_k, v| v.nil? || v.is_a?(Array) && v.empty? }
+        }.delete_if{ |_k, v|
+          v.nil? || v.is_a?(Array) && v.empty?
+        }
+      }
```

Comment:
I think if we extract the logic here into functions `collect_X(vrp)` for X in `vehicles`, `jobs` and `shipments` `run_vroom` would stay more legible. Maybe matrix logic as well if you think it is easy to extract.

```ruby
problem[:vehicles] = collect_vehicles(vrp)
problem[:jobs] = collect_jobs(vrp)
problem[:shipments] = collect_shipments(vrp)
```